File: sources/sdk/java/src/main/java/io/drasi/source/sdk/models/SourceElement.java
/*
 * Copyright 2024 The Drasi Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.drasi.source.sdk.models;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import java.util.Set;

/**
 * Represents a single element in the queryable graph, either a node or relation.
 */
public class SourceElement {

    private String id;
    private JsonNode properties;
    private Set<String> labels;
    private String startId;
    private String endId;


    /**
     * Create a new Node with the given id, properties and labels.
     *
     * @param id         The id of the element.
     * @param properties The properties of the element.
     * @param labels     The labels of the element.
     */
    public SourceElement(String id, JsonNode properties, Set<String> labels) {
        this.id = id;
        this.properties = properties;
        this.labels = labels;
    }

    /**
     * Create a new Relation with the given id, properties, labels, startId and endId.
     * The startId and endId are the ids of the nodes that the relation connects.
     *
     * @param id         The id of the element.
     * @param properties The properties of the element.
     * @param labels     The labels of the element.
     * @param startId    The id of the start node.
     * @param endId      The id of the end node.
     *
     */
    public SourceElement(String id, JsonNode properties, Set<String> labels, String startId, String endId) {
        this.id = id;
        this.properties = properties;
        this.labels = labels;
        this.startId = startId;
        this.endId = endId;
    }

    public String toJson() {
        var result = ConvertProperties();
        return result.toString();
    }

    private JsonNode ConvertProperties() {
        var result = JsonNodeFactory.instance.objectNode();

        result.put("id", id);
        var lbls = JsonNodeFactory.instance.arrayNode();
        for (var lbl : labels)
            lbls.add(lbl);
        result.set("labels", lbls);

        if (properties != null) {
            var props = JsonNodeFactory.instance.objectNode();
            var pgFields = properties.fields();
            while (pgFields.hasNext()) {
                var field = pgFields.next();
                props.set(field.getKey(), field.getValue());
            }
            result.set("properties", properties);
        }

        if (startId != null) {
            result.put("start_id", startId);
        }

        if (endId != null) {
            result.put("end_id", endId);
        }

        return  result;
    }

    enum Op {
        INSERT,
        UPDATE,
        DELETE
    }
}

File: sources/sdk/dotnet/Drasi.Source.SDK/Models/SourceElement.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Drasi.Source.SDK.Models;

public class SourceElement
{
    private readonly string id;
    private readonly JsonObject? properties;
    private readonly HashSet<string> labels;
    private readonly string? startId;
    private readonly string? endId;

    public ElementType Type { get; private set; }

    /// <summary>
    /// Create a new SourceElement that represents a node
    /// </summary>
    /// <param name="id">Unique ID of the node within the context of the entire source</param>
    /// <param name="labels">Labels for the node in the property graph</param>
    /// <param name="properties">Properties of the node</param>
    public SourceElement(string id, HashSet<string> labels, JsonObject? properties)
    {
        this.id = id;
        this.properties = properties;
        this.labels = labels;
        this.Type = ElementType.Node;
    }

    /// <summary>
    /// Create a new SourceElement that represents a relation
    /// </summary>
    /// <param name="id">Unique ID of the relation within the context of the entire source</param>
    /// <param name="labels">Labels for the relation in the property graph</param>
    /// <param name="properties">Properties of the relation</param>
    /// <param name="startId">ID of the inbound node for this relationship</param>
    /// <param name="endId">ID of the outbound  node for this relationship</param>
    public SourceElement(string id, HashSet<string> labels, JsonObject? properties, string startId, string endId)
    {
        this.id = id;
        this.properties = properties;
        this.labels = labels;
        this.startId = startId;
        this.endId = endId;
        this.Type = ElementType.Relation;
    }

    public string ToJson()
    {
        var result = ToJsonObject();
        return result.ToJsonString();
    }

    public JsonObject ToJsonObject()
    {
        var result = new JsonObject
        {
            { "id", id }
        };

        var lbls = new JsonArray();
        foreach (var lbl in labels)
        {
            lbls.Add(lbl);
        }
        result.Add("labels", lbls);

        if (properties != null)
        {
            result.Add("properties", properties);
        }

        if (startId != null)
        {
            result.Add("startId", startId);
        }

        if (endId != null)
        {
            result.Add("endId", endId);
        }

        return result;
    }

    public enum ElementType
    {
        Node,
        Relation
    }
}